tools/gfx/render-d3d12.cpp

```diff
@@ -1278,82 +1288,24 @@ Result D3D12Renderer::_bindRenderState(PipelineStateImpl* pipelineStateImpl, ID3
 
 // !!!!!!!!!!!!!!!!!!!!!!!!!!!! Renderer interface !!!!!!!!!!!!!!!!!!!!!!!!!!
 
-Result D3D12Renderer::initialize(const Desc& desc, void* inWindowHandle)
-{
-    m_hwnd = (HWND)inWindowHandle;
-    // Rather than statically link against D3D, we load it dynamically.
 
-    HMODULE d3dModule = LoadLibraryA("d3d12.dll");
-    if (!d3dModule)
-    {
-        fprintf(stderr, "error: failed load 'd3d12.dll'\n");
-        return SLANG_FAIL;
-    }
-
-    HMODULE dxgiModule = LoadLibraryA("Dxgi.dll");
-    if (!dxgiModule)
-    {
-        fprintf(stderr, "error: failed load 'dxgi.dll'\n");
-        return SLANG_FAIL;
-    }
-
-
-#define LOAD_D3D_PROC(TYPE, NAME) \
-        TYPE NAME##_ = (TYPE) loadProc(d3dModule, #NAME);
-#define LOAD_DXGI_PROC(TYPE, NAME) \
-        TYPE NAME##_ = (TYPE) loadProc(dxgiModule, #NAME);
-
-    UINT dxgiFactoryFlags = 0;
-
-#if ENABLE_DEBUG_LAYER
-    {
-        LOAD_D3D_PROC(PFN_D3D12_GET_DEBUG_INTERFACE, D3D12GetDebugInterface);
-        if (D3D12GetDebugInterface_)
-        {
-            if (SUCCEEDED(D3D12GetDebugInterface_(IID_PPV_ARGS(m_dxDebug.writeRef()))))
-            {
-                m_dxDebug->EnableDebugLayer();
-                dxgiFactoryFlags |= DXGI_CREATE_FACTORY_DEBUG;
-            }
-        }
-    }
-#endif
-
-    m_D3D12SerializeRootSignature = (PFN_D3D12_SERIALIZE_ROOT_SIGNATURE)loadProc(d3dModule, "D3D12SerializeRootSignature");
-    if (!m_D3D12SerializeRootSignature)
-    {
-        return SLANG_FAIL;
-    }
+Result D3D12Renderer::_createAdaptor(DeviceCheckFlags deviceCheckFlags, ComPtr<IDXGIFactory4>& outDxgiFactory, ComPtr<IDXGIAdapter>& outAdapter)
+{
+    const UINT dxgiFactoryFlags = (deviceCheckFlags & DeviceCheckFlag::UseDebug) ? DXGI_CREATE_FACTORY_DEBUG  : 0;
 
     // Try and create DXGIFactory
     ComPtr<IDXGIFactory4> dxgiFactory;
-    {
-        typedef HRESULT(WINAPI *PFN_DXGI_CREATE_FACTORY_2)(UINT Flags, REFIID riid, _COM_Outptr_ void **ppFactory);
-        LOAD_DXGI_PROC(PFN_DXGI_CREATE_FACTORY_2, CreateDXGIFactory2);
-        if (!CreateDXGIFactory2_)
-        {
-            return SLANG_FAIL;
-        }
-        SLANG_RETURN_ON_FAIL(CreateDXGIFactory2_(dxgiFactoryFlags, IID_PPV_ARGS(dxgiFactory.writeRef())));
-    }
-
-    D3D_FEATURE_LEVEL featureLevel = D3D_FEATURE_LEVEL_11_0;
+    SLANG_RETURN_ON_FAIL(m_CreateDXGIFactory2(dxgiFactoryFlags, IID_PPV_ARGS(dxgiFactory.writeRef())));
+    
+    D3D_FEATURE_LEVEL featureLevel = DeviceCheckFlag::UseFullFeatureLevel ? D3D_FEATURE_LEVEL_11_1 : D3D_FEATURE_LEVEL_11_0;
```

Comment:
I think that we don't need/want to do this axis of variation in the D3D12 case.

My understanding from the MSDN documentation for `D3D12CreateDevice` is that the feature level is a *minimum* feature level, so that device creation will only fail if you can't get at least that level, but the device you get back will be able to report an actual supported feature level that might be higher.

For the purposes of the testing app I think we can always just ask for `D3D_FEATURE_LEVEL_11_0`.

I guess the only complication there is that you are adding the flag for "use full feature level" before the flag for "use hardware device," which (IIUC) means that we will pick a warp device that supports 11.1 over a hardware device that supports 11.0 with the current code, while my proposed change would switch that around.

I honestly don't know what we want the defaults to be, but it seems like we should favor running on hardware whenever possible, no?

Reply:
It iterates over the flags in the order specified by add so it will try 11_1, 11_0 on hardware, then the same of reference, and then with release. 

That the change tries to make device startup work similarly to Dx11 (not least because it fixes problem with not being able to startup with debug). 

So it does prefer hardware (trying hardware on 11_1 and 11_0 first). But if they both fail, then tries reference... and if they both fail tries debug on all these combinations.

Reply:
So the code has:

```c++
FlagCombiner combiner;
combiner.add(DeviceCheckFlag::UseFullFeatureLevel, ChangeType::OnOff);
combiner.add(DeviceCheckFlag::UseHardwareDevice, ChangeType::OnOff);
```

I would have assumed that the flags that get added first represent the "outer loop" when enumerating combinations, and then the flags that get added later are "inner loops." So I'd expected something like:

```
for(auto useFullFeature : {true, false}) {
    for(auto useHardwareDevice : {true, false}) {
        tryToCreateDevice();
    }
}
```

And my read on that means that we'd try:
* full feature, hardware
* full feature, no hardware
* no full feature, hardware,
* no full feature, no hardware

But your response says:

> So it does prefer hardware (trying hardware on 11_1 and 11_0 first).

Am I wrong about the order of enumeration, or missing something else?

Reply:
"I would have assumed that the flags that get added first represent the "outer loop" when enumerating combinations, and then the flags that get added later are "inner loops.""

The order is the opposite - that first added is iterated over at the highest frequency. I wrote the class to try and make the logic around the testing clearer, but perhaps that didn't work out. Would be relatively easy to make it so it went in the nested for order you propose, and perhaps that's a clearer way.